c/parallel/test/algorithm_execution.h

```diff
@@ -0,0 +1,174 @@
+  }
+
+  int get_cc_major() const
+  {
+    return cc_major;
+  }
+  int get_cc_minor() const
+  {
+    return cc_minor;
+  }
+  const char* get_cub_path() const
+  {
+    return cub_path;
+  }
+  const char* get_thrust_path() const
+  {
+    return thrust_path;
+  }
+  const char* get_libcudacxx_path() const
+  {
+    return libcudacxx_path;
+  }
+  const char* get_ctk_path() const
+  {
+    return ctk_path;
+  }
+};
+
+template <typename BuildResultT,
+          typename Build,
+          typename Cleanup,
+          typename Run,
+          typename BuildCache,
+          typename KeyT,
+          typename... Tx>
+void AlgorithmExecute(std::optional<BuildCache>& cache, const std::optional<KeyT>& lookup_key, Tx&&... args)
+{
+  constexpr int device_id = 0;
+  const auto& build_info  = BuildInformation<device_id>::init();
+
+  BuildResultT build;
+
+  bool found               = false;
+  const bool cache_and_key = bool(cache) && bool(lookup_key);
+
+  if (cache_and_key)
+  {
+    auto& cache_v     = cache.value();
+    const auto& key_v = lookup_key.value();
+    if (cache_v.contains(key_v))
+    {
+      build = cache_v.get(key_v).get();
+      found = true;
+    }
+  }
+
+  if (!found)
+  {
+    REQUIRE(
+      CUDA_SUCCESS
+      == Build{}(&build,
+                 args...,
+                 build_info.get_cc_major(),
+                 build_info.get_cc_minor(),
+                 build_info.get_cub_path(),
+                 build_info.get_thrust_path(),
+                 build_info.get_libcudacxx_path(),
+                 build_info.get_ctk_path()));
+
+    if (cache_and_key)
+    {
+      auto& cache_v     = cache.value();
+      const auto& key_v = lookup_key.value();
+      cache_v.insert(key_v, build);
+    }
+  }
+
+  const std::string& sass = inspect_sass(build.cubin, build.cubin_size);
+
+  REQUIRE(sass.find("LDL") == std::string::npos);
+  REQUIRE(sass.find("STL") == std::string::npos);
+
+  CUstream null_stream = 0;
+
+  size_t temp_storage_bytes = 0;
+  REQUIRE(CUDA_SUCCESS == Run{}(build, nullptr, &temp_storage_bytes, args..., null_stream));
+
+  pointer_t<uint8_t> temp_storage(temp_storage_bytes);
+
+  REQUIRE(CUDA_SUCCESS == Run{}(build, temp_storage.ptr, &temp_storage_bytes, args..., null_stream));
+
+  if (cache_and_key)
+  {
+    // if cache and lookup_key were provided, the ownership of resources
+    // allocated for build is transferred to the cache, hence do nothing
+  }
+  else
+  {
+    // release build data resources
+    REQUIRE(CUDA_SUCCESS == Cleanup{}(&build));
+  }
+}
+
+template <typename BuildResultT, typename Cleanup>
+struct BuildResultDeleter
+{
+  static constexpr Cleanup cleanup_{};
+  void operator()(BuildResultT* build_data) const noexcept
+  {
+    check_success(cleanup_(build_data));
+  }
+
+private:
+  void check_success(CUresult status) const noexcept
+  {
+    if (status != CUDA_SUCCESS)
+    {
+      std::cerr << "Clean-up call returned status " << status << std::endl;
+    }
+  }
```

Comment:
Why is this a separate method?

Reply:
Mostly for readability. To address performance concern (no need to pass `this`), I am going to change this private method to `static`.